Sources/JS/initTypeCollection.js

```diff
@@ -24,50 +32,43 @@ Type.prototype.union = function(otherType) {
     return newType
 }
 Type.prototype.setGroup = function(obj) {
-    for (var i=0;i<possibleGroups.length;i++) {
+    for (var i=0;i<orderedGroups.length;i++) {
         try {
-            if (possibleGroups[i].belongsToGroup(obj)) {
-                this.group = possibleGroups[i].name
-                return
+            if (orderedGroups[i].belongsToGroup(obj)) {
+                this.group = orderedGroups[i].name
+                return orderedGroups[i]
             }
         } catch(err) {}
     }
 }
 
 // variableNumber: Type
 var types = {}
-function getBaseType(rawValue) {
-    var newType = new Type()
-    newType.definiteType = rawValue
-    newType.possibleType = rawValue
-    if (rawValue === baseTypes.object) {
-        newType.properties = []
-        newType.methods = []
-    }
-    return newType
-}
 function getCurrentType(value){
     try {
         // Fuzzilli handles both null/undefined as undefined
-        if (value == null) return getBaseType(baseTypes.undefined)
+        if (value == null) return new Type(baseTypes.undefined)
         try {
-            if (isInteger(value)) return getBaseType(baseTypes.integer)
+            if (isInteger(value)) return new Type(baseTypes.integer)
         } catch(err) {}
-        if (typeof value === 'number') return getBaseType(baseTypes.float)
-        if (typeof value === 'string') return getBaseType(baseTypes.string)
-        if (typeof value === 'boolean') return getBaseType(baseTypes.boolean)
-        if (typeof value === 'bigint') return getBaseType(baseTypes.bigint)
+        if (typeof value === 'number') return new Type(baseTypes.float)
+        if (typeof value === 'string') return new Type(baseTypes.string)
+        if (typeof value === 'boolean') return new Type(baseTypes.boolean)
+        if (typeof value === 'bigint') return new Type(baseTypes.bigint)
         try {
-            if (value instanceof RegExp) return getBaseType(baseTypes.regexp)
+            if (value instanceof RegExp) return new Type(baseTypes.regexp)
         } catch(err) {}
         if (typeof value === 'object' || typeof value === 'symbol') {
-            var currentType = getBaseType(baseTypes.object)
-            currentType.setGroup(value)
+            var currentType = new Type(baseTypes.object)
+            var group = currentType.setGroup(value)
+            // handle long arrays specially to reduce time spent on properties collection
+            if (group && group.slowTypeCollection && value.length > maxArrayLength) {
```

Comment:
Should this check happen before or inside the `while` loop since you could have something like `Object.prototype.__proto__ = new Uint8Array(10000)` which would not be handled by the current logic? But maybe that edge case happens so rarely that it's not worth bothering

Reply:
yeah I thought about that, but it would mean we have to check group for every object in prototype chain which increases number of instructions needed to be done for every object from original code
and I did not encounter such case, so maybe just not worth it?

Reply:
SGTM!